### DISCUSSION TITLE
Why don't StreamProvider listeners get cancelled upon hot reload?

### DISCUSSION THREAD

**Discussion by plammens** [rrousselGit/riverpod#1911] [Q&A]
I use `StreamProvider`s to listen to Firestore data.
When I hot reload my app, it seems the stream subscriptions don't get properly cancelled, i.e. it seems I end up with a fresh batch of subscriptions together with the old subscriptions which aren't used by any active widget instance anymore.
This happens even with `autoDispose` activated.

Here is a reproducible example, it's the most minimal I managed to come up with.

1. Create a new Flutter project with `flutter create`
2. Initialise firebase with `flutterfire config` ([detailed instructions](https://firebase.google.com/docs/flutter/setup))
3. (Optional) Set up Firebase emulator suite ([detailed instructions](https://firebase.google.com/docs/emulator-suite/install_and_configure))
4. Copy this in `main.dart`:
    Flutter app:
    ```dart
    import 'package:cloud_firestore/cloud_firestore.dart';
    import 'package:firebase_core/firebase_core.dart';
    import 'package:flutter/material.dart';
    import 'package:flutter_riverpod/flutter_riverpod.dart';
    import 'package:sandbox/firebase_options.dart';
    
    final provider = StreamProvider.autoDispose((ref) =>
        FirebaseFirestore.instance.collection("users").snapshots().map((event) {
          print("Listener got stream event: $event");
          return event;
        }));
    
    void main() async {
      WidgetsFlutterBinding.ensureInitialized();
      await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);
      FirebaseFirestore.instance.useFirestoreEmulator("localhost", 8080);  // optional
      runApp(const ProviderScope(child: MyApp()));
    }
    
    class MyApp extends StatelessWidget {
      const MyApp({super.key});
    
      @override
      Widget build(BuildContext context) {
        return const MaterialApp(title: 'Flutter Demo', home: MyHomePage());
      }
    }
    
    class MyHomePage extends ConsumerWidget {
      const MyHomePage({super.key});
    
      @override
      Widget build(BuildContext context, WidgetRef ref) {
        var value = ref.watch(provider);
        return Text(value.toString());
      }
    }
    
    ```
5. (Optional) Start up Firestore emulator
6. Create a dummy collection inside Firestore, in this case I created one named "users" with one dummy document
7. Start the flutter app. **In the console, one line like "Listener got stream event: ..." appears.**
8. Hot reload the flutter app.
9. Modify the Firestore collection, e.g. by adding a document. **In the console, the line "Listener got stream event: ..." appears twice, indicating that there are two active stream subscriptions.**

**Comment by rrousselGit** [comment]
Which platform are you on? Flutter web? I believe there's an issue with Flutter web's hot-reload (well hot-restart for that matter) not properly clearly streams

**Reply by plammens** [reply]
Yes, I'm mainly testing on Flutter web with Chrome on Windows.

**Reply by rrousselGit** [reply]
That's likely the problem then
